### PR TITLE
(virtualbox) GH-423

### PR DESCRIPTION
**Pull request by majkinetor**
https://github.com/chocolatey/chocolatey-coreteampackages/issues/423

### CODE REVIEW

File: automatic/virtualbox/tools/chocolateyInstall.ps1 [hunk 1]
```diff
@@ -1,9 +1,20 @@
...
+
 $cert = ls cert: -Recurse | ? { $_.Thumbprint -eq 'a88fd9bdaa06bc0f3c491ba51e231be35f8d1ad5' }
 if (!$cert) {
     $toolsPath = Split-Path $MyInvocation.MyCommand.Definition
     Start-ChocolateyProcessAsAdmin "certutil -addstore 'TrustedPublisher' '$toolsPath\oracle.cer'"
 }
 
+$pp = Get-PackageParameters
+$silentArgs = @('-s -l -msiparams REBOOT=ReallySuppress')
+$silentArgs += if (!$pp.CurrentUser) { 'ALLUSERS=1' } else { 'ALLUSERS=2' Write-Host 'Param: Installing for current user' }
```
Comment (AdmiringWorm) [inline automatic/virtualbox/tools/chocolateyInstall.ps1]: Missing semicolon (`;`) between *'ALLUSERS=2* and `Write-Host`?

Comment (majkinetor) [inline automatic/virtualbox/tools/chocolateyInstall.ps1]: Fixed.

File is included.

**Comment by AdmiringWorm** [comment]
I'm still getting errors when trying to run the beforemodify script.
Something regarding that it cannot bind argument to parameter 'Path' because it is null.

Not sure where it fails yet.
I think it is because it can't locate my virtualbox installation (don't have it at a default location),
and Test-Path can't handle nulls?

**Comment by majkinetor** [comment]
ARGH... another typo, had to add ( ... sorry about this, was doing that part in a hurry. 
Tested it up now, definitelly should work.

**Comment by AdmiringWorm** [comment]
the before modify scripts runs successfully (well without error at least).
the script as is will probably only work for users that have virtualbox installed in the default location.

would it perhaps be possible to fall back on one of the following keys:
```
[HKEY_CLASSES_ROOT\progId_VirtualBox.Shell.hdd\DefaultIcon]
@="\"D:\\ChocoProgs\\VirtualBox\\VBoxRes.dll\",-306"
[HKEY_CLASSES_ROOT\progId_VirtualBox.Shell.ova\DefaultIcon]
@="\"D:\\ChocoProgs\\VirtualBox\\VBoxRes.dll\",-302"
[HKEY_CLASSES_ROOT\progId_VirtualBox.Shell.ova\shell\open\command]
@="\"D:\\ChocoProgs\\VirtualBox\\VirtualBox.exe\" \"%1\""
[HKEY_CLASSES_ROOT\progId_VirtualBox.Shell.vbox\DefaultIcon]
@="\"D:\\ChocoProgs\\VirtualBox\\VBoxRes.dll\",-201"
[HKEY_CLASSES_ROOT\progId_VirtualBox.Shell.vbox\shell\open\command]
@="\"D:\\ChocoProgs\\VirtualBox\\VirtualBox.exe\" \"%1\""
[HKEY_CLASSES_ROOT\TypeLib\{D7569351-1750-46F0-936E-BD127D5BC264}\1.3\HELPDIR]
@="D:\\ChocoProgs\\VirtualBox\\"
[HKEY_CLASSES_ROOT\WOW6432Node\TypeLib\{D7569351-1750-46F0-936E-BD127D5BC264}\1.3\HELPDIR]
@="D:\\ChocoProgs\\VirtualBox\\"
```

so it will work for users who changed directory (like me)

**Comment by majkinetor** [comment]
> The script as is will probably only work for users that have virtualbox installed in the default location.

No, it should ll be found by application registration (virtualbox.exe is registred by the installer). I am generally hesistant to add anything here that is app specific and can't be added to `Get-AppInstallLocation`. 

Abstracting above probably is doable only using full registry search whcih is similar to full file system search.  Both could be done with tools quickly tho.

**Comment by majkinetor** [comment]
> The before modify scripts runs successfully (well without error at least).

It should do nothing if vbox is not found. It should show output only if machine is running.

**Comment by majkinetor** [comment]
I would add option `/InstallExtensionPack` but will not do that until choco can rebember the passed parameters.

So I will add now `/NoExtensionPack` to do the oposite.

**Comment by AdmiringWorm** [comment]
sound good

**Comment by ferventcoder** [comment]
Extension pack with this is great!

**Comment by ferventcoder** [comment]
Thanks @majkinetor !!

**Comment by majkinetor** [comment]
One thing tho confuses me:

```
Successfully installed "Oracle VM VirtualBox Extension Pack".
Adding to PATH if needed
virtualbox installed to 'C:\Program Files\Oracle\VirtualBox'
virtualbox registered as vbox
Only an exit code of non-zero will fail the package by default. Set
 `--failonstderr` if you want error messages to also fail a script. See
 `choco -h` for details.
 The install of virtualbox was successful.
  Software installed as 'EXE', install location is likely default.
```

Message 'Only an exit code of non-zero will fail ....'

**Comment by ferventcoder** [comment]
Last comment deleted?

Non-zero exit codes message - just means that something in the install wrote to stderr. Chocolatey won't fail because of that. Because there are crappy tools like reg.exe that always write to stderr instead of stdout.

**Comment by AdmiringWorm** [comment]
@majkinetor I just tried to test this package, but got an error almost directly after the installation of virtualbox itself:

Relevant parts of the log:
```
2016-11-28 23:52:26,221 [DEBUG] - Finishing 'Start-ChocolateyProcessAsAdmin'
2016-11-28 23:52:26,226 [INFO ] - virtualbox has been installed.
2016-11-28 23:52:26,242 [INFO ] - VERBOSE: Trying local and machine (x32 & x64) Uninstall keys
2016-11-28 23:52:26,248 [DEBUG] - Running 'Get-UninstallRegistryKey' for 'virtualbox' with SoftwareName:'virtualbox'
2016-11-28 23:52:26,251 [INFO ] - VERBOSE: Retrieving all uninstall registry keys
2016-11-28 23:52:26,842 [DEBUG] - Registry uninstall keys on system: 638
2016-11-28 23:52:26,845 [DEBUG] - Error handling check: 'Get-ItemProperty' fails if a registry key is encoded incorrectly.
2016-11-28 23:52:27,245 [DEBUG] - Found 0 uninstall registry key(s) with SoftwareName:'virtualbox'
2016-11-28 23:52:27,249 [INFO ] - VERBOSE: Found 0 keys, aborting this method
2016-11-28 23:52:27,254 [DEBUG] - Running Get-ProcessorBits -compare '64' 
2016-11-28 23:52:27,257 [INFO ] - VERBOSE: Trying Program Files with 2 levels depth: C:\Program Files C:\Program Files\*\* C:\Program Files (x86) C:\Program Files (x86)\*\*
2016-11-28 23:52:27,481 [INFO ] - VERBOSE: Trying native commands on PATH
2016-11-28 23:52:27,897 [INFO ] - VERBOSE: Trying Registry: \SOFTWARE\Microsoft\Windows\CurrentVersion\App Paths
2016-11-28 23:52:27,929 [INFO ] - VERBOSE: No location found
2016-11-28 23:52:28,125 [ERROR] - ERROR: Cannot bind argument to parameter 'Path' because it is null.
```

**Comment by majkinetor** [comment]
Fixed that worm, thx.

Judging from *VERBOSE: No location found* you installed to custom folder ?

**Comment by AdmiringWorm** [comment]
yeah, that's correct.
whenever it's possible I install programs to a separate partition.

File: automatic/virtualbox/tools/chocolateyBeforeModify.ps1 [hunk 2]
```diff
@@ -0,0 +1,35 @@
+$packageName = 'virtualbox'
+$pvbox = ps virtualbox -ea 0
+if (!$pvbox) {
+    Write-Host "$packageName is not running"
+    return
+}
+
+$installLocation = Get-AppInstallLocation $packageName
+if (!$installLocation) { $installLocation = Split-Path (gcm VBoxManage.exe -ea 0).Path }
```
Comment (majkinetor) [inline automatic/virtualbox/tools/chocolateyBeforeModify.ps1]: you are right. thx.

**Comment by majkinetor** [comment]
OK, I got the idea.

`Get-AppInstallLocation` can't find it in that case as MSI registry record doesn't contain the path. To get it, I would have to use MSI API function. 

I will try to update the `core.extension` so that above mentioned function can handle this.

**Comment by majkinetor** [comment]
Have you tried it with running virtualbox and 1 of its machines prior to update ?

**Comment by majkinetor** [comment]
If you installed it to custom location I think it didn't install extension for you ...

**Comment by AdmiringWorm** [comment]
> Have you tried it with running virtualbox and 1 of its machines prior to update ?

yeah, I tried upgrading it with 1 vm running. and it closes it, or rather aborts it.

> If you installed it to custom location I think it didn't install extension for you ...

Before testing it, I removed virtualbox and installed it to the default.
But now that you mention it, I checked under extensions on virtualbox and nothing is showing up!

**Comment by majkinetor** [comment]
> yeah, I tried upgrading it with 1 vm running. and it closes it, or rather aborts it.

Mhm... it should list running machines. It looks like `vboxmanage list runningvms` doesn't work there too. I kill virtualbox only after I do this.

> But now that you mention it, I checked under extensions on virtualbox and nothing is showing up!

The installation output should include `virtualbox-extension` section. This is my output:

```
## virtualbox~> Test-Package -Vagrant $null

Package info
  Path:         C:\work\_chocolatey\chocolatey-coreteampackages\automatic\virtualbox\virtualbox.5.1.10.nupkg
  Name:         virtualbox
  Version:      5.1.10

Testing package install
Installing the following packages:
virtualbox
By installing you accept licenses for the packages.

virtualbox v5.1.10 (forced)
virtualbox package files install completed. Performing other installation steps.
Downloading virtualbox 64 bit
  from 'http://download.virtualbox.org/virtualbox/5.1.10/VirtualBox-5.1.10-112026-Win.exe'
Download of VirtualBox-5.1.10-112026-Win.exe (117.32 MB) completed.
Hashes match.
Installing virtualbox...
virtualbox has been installed.

Installing extension pack
Downloading virtualbox-extensionpack 64 bit
  from 'http://download.virtualbox.org/virtualbox/5.1.10/Oracle_VM_VirtualBox_Extension_Pack-5.1.10.vbox-extpack'
WARNING: C:\Users\majkinetor\AppData\Local\Temp\chocolatey\virtualbox\5.1.10\Oracle_VM_VirtualBox_Extension_Pack-5.1.10.vbox-extpack is of content type text/plain
Download of Oracle_VM_VirtualBox_Extension_Pack-5.1.10.vbox-extpack (18.57 MB) completed.
Hashes match.
C:\Users\majkinetor\AppData\Local\Temp\chocolatey\virtualbox\5.1.10\Oracle_VM_VirtualBox_Extension_Pack-5.1.10.vbox-extpack
0%...10%...20%...30%...40%...50%...60%...70%...80%...90%...100%
Successfully installed "Oracle VM VirtualBox Extension Pack".
Adding to PATH if needed
virtualbox installed to 'C:\Program Files\Oracle\VirtualBox'
virtualbox registered as vbox
 The install of virtualbox was successful.
  Software installed as 'EXE', install location is likely default.
```

**Comment by majkinetor** [comment]
Next run with `cup virtualbox -source . --force` and 1 running machine also has this one:

```
virtualbox v5.1.10 is the latest version available based on your source(s).
Virtualbox is running, killing it prior to update
```

however, extension fails when vbox is killed:

```
WARNING: C:\Users\majkinetor\AppData\Local\Temp\chocolatey\virtualbox\5.1.10\Oracle_VM_VirtualBox_Extension_Pack-5.1.10.vbox-extpack is of content type text/plain
Progress: 100% - Completed download of C:\Users\majkinetor\AppData\Local\Temp\chocolatey\virtualbox\5.1.10\Oracle_VM_VirtualBox_Extension_Pack-5.1.10.vbox-extpack (18.57 MB).
Download of Oracle_VM_VirtualBox_Extension_Pack-5.1.10.vbox-extpack (18.57 MB) completed.
Hashes match.
C:\Users\majkinetor\AppData\Local\Temp\chocolatey\virtualbox\5.1.10\Oracle_VM_VirtualBox_Extension_Pack-5.1.10.vbox-extpack
0%...
Progress state: E_FAIL

VBoxManage.exe: error: Failed to install "C:\Users\majkinetor\AppData\Local\Temp\chocolatey\virtualbox\5.1.10\Oracle_VM_VirtualBox_Extension_Pack-5.1.10.vbox-extpack"
VBoxManage.exe: error: The installer failed with exit code 1: VBoxExtPackHelperApp.exe: error: Failed to delete the extension pack directory: VERR_ACCESS_DENIED ('C:\Program Files\Oracle\VirtualBox\ExtensionPacks\Oracle_VM_VirtualBox_Extension_Pack-_-uninst')
VBoxManage.exe: error: rcExit=1
VBoxManage.exe: error: Details: code E_FAIL (0x80004005), component ExtPackManagerWrap, interface IExtPackManager
VBoxManage.exe: error: Context: "enum RTEXITCODE __cdecl handleExtPack(struct HandlerArg *)" at line 1158 of file VBoxManageMisc.cpp
```

so it doesn't list machines

**Comment by majkinetor** [comment]
The directory remains locked. Nothing locks it based on LockHunter and I must restart computer for installation to work again.

**Comment by majkinetor** [comment]
This doesn't help too:

```
## virtualbox~> VBoxManage extpack cleanup
VBoxManage.exe: error: The installer failed with exit code 1: VBoxExtPackHelperApp.exe: error: Failed to delete the temporary extension pack directory: VERR_ACCESS_DENIED ('C:\Program Files\Oracle\VirtualBox\ExtensionPacks\Oracle_VM_VirtualBox_Extension_Pack-_-uninst')
VBoxManage.exe: error: rcExit=1
VBoxManage.exe: error: Details: code E_FAIL (0x80004005), component ExtPackManagerWrap, interface IExtPackManager, callee IUnknown
VBoxManage.exe: error: Context: "Cleanup()" at line 1208 of file VBoxManageMisc.cpp
```

**Comment by AdmiringWorm** [comment]
regarding the `vboxmanage list`
it looks like it must be run as the current user, not in a elevated prompt.

Can't say I really noticed the error, but it could be it just escaped my attention

**Comment by majkinetor** [comment]
> it looks like it must be run as the current user, not in a elevated prompt.

Good find !!!

> Can't say I really noticed the error, but it could be it just escaped my attention

Its intentional that package passes without ext becuase its not big enough failure to cause entire vbox package to fail.

**Comment by majkinetor** [comment]
@ferventcoder @gep13 

Is there any well known method for situations like this ? We need to execute non-elevated code from elevated session.

**Comment by majkinetor** [comment]
I posted the question on [stackoverflow](http://stackoverflow.com/questions/40863475/starting-non-elevated-prompt-from-elevated-session?noredirect=1#comment68942700_40863475).

**Comment by majkinetor** [comment]
I created function [Start-ProcessNonEleveated.ps1](https://gist.github.com/majkinetor/7e9ca2246a6cfc320aeeaab392d5747d) that looks like working and its using Task Scheduler.

It returns HashTable having `err` and `out` object.

```
Start-ProcessNonElevated 'vboxmanage list runningvms' -UsePowerShell
Name                           Value
----                           -----
err
out                            "chocolatey-test-environment_default_1480354084003_60014" {31409fff-c195-4a83-ab12-96bba020e051}


PS> Start-ProcessNonElevated 'vboxmanage dummy' -UsePowerShell | % err
vboxmanage : Oracle VM VirtualBox Command Line Management Interface Version 5.1.10
At C:\Users\majkinetor\AppData\Local\Temp\tmp76D0.ps1:2 char:1
+ vboxmanage dummy >"$r.out.log" 2>"$r.err.log"
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (Oracle VM Virtu... Version 5.1.10:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError

(C) 2005-2016 Oracle Corporation
All rights reserved.

Usage:

  VBoxManage [<general option>] <command>
```

**Comment by majkinetor** [comment]
I finished non elevated code for shutdown of machines. It looks like its working OK up to the point where some machines do not respond to `acpipowerbutton` - windows server has it disabled by default and on the internet everybody mentions GPO setting

I tried this on chocolatey test environment without success.

```powershell
#For acpipowerbutton to work on Windows server see: http://ethertubes.com/unattended-acpi-shutdown-of-windows-server/
Write-Host 'Enable acpi shutdown'
sp HKLM\SOFTWARE\Microsoft\Windows\CurrentVersion\policies\system shutdownwithoutlogon 1
gpupdate /force
```

Using the `gpedit.msc` and rebooting doesn't make the differece either. The function doesn't work using the VirtualBox GUI (Machine->Close ->ACPI Shutdown or CTRL+H).

For this reason I added this `hardcoded option` to [BeforeModify](https://github.com/chocolatey/chocolatey-coreteampackages/blob/8d3e8e0d63450ff8f63193ae14d767d85154e3dc/automatic/virtualbox/tools/chocolateyBeforeModify.ps1#L8).

```
$shutdown_type = 'acpipowerbutton'  # 'acpipowerbutton', 'poweroff', 'savestate'
```

- `Poweroff` works always but machine may enter recovery mode after it
- `savestate` works always but requires a lot of memory


It turns out that this request is more problematic then originally anticipated @ferventcoder. Hower, I think that this code wont harm to stay. I test it with:

```
cpack; choco install virtualbox -source . --skip-scripts --force; cup virtualbox -source . --force
```

from within virtualbox package directory.

**Comment by majkinetor** [comment]
I think we should use `savestate`.

This is the output from my system with choco test env working:

```
By upgrading you accept licenses for the packages.
virtualbox v5.1.10 is the latest version available based on your source(s).
virtualbox is running, trying to gracefully shutdown any running machines
Number of machines running: 1
Machine "chocolatey-test-environment_default_1480354084003_60014" powered down prior to installation using 'savestate' method

virtualbox v5.1.10 (forced)
virtualbox package files upgrade completed. Performing other installation steps.
Downloading virtualbox 64 bit
  from 'http://download.virtualbox.org/virtualbox/5.1.10/VirtualBox-5.1.10-112026-Win.exe'
Progress: 100% - Completed download of C:\Users\majkinetor\AppData\Local\Temp\chocolatey\virtualbox\5.1.10\VirtualBox-5.1.10-112026-Win.exe (117.32 MB).
Download of VirtualBox-5.1.10-112026-Win.exe (117.32 MB) completed.
Hashes match.
Installing virtualbox...
virtualbox has been installed.

Installing extension pack
Downloading virtualbox-extensionpack 64 bit
  from 'http://download.virtualbox.org/virtualbox/5.1.10/Oracle_VM_VirtualBox_Extension_Pack-5.1.10.vbox-extpack'
WARNING: C:\Users\majkinetor\AppData\Local\Temp\chocolatey\virtualbox\5.1.10\Oracle_VM_VirtualBox_Extension_Pack-5.1.10.vbox-extpack is of content type text/plain
Progress: 100% - Completed download of C:\Users\majkinetor\AppData\Local\Temp\chocolatey\virtualbox\5.1.10\Oracle_VM_VirtualBox_Extension_Pack-5.1.10.vbox-extpack (18.57 MB).
Download of Oracle_VM_VirtualBox_Extension_Pack-5.1.10.vbox-extpack (18.57 MB) completed.
Hashes match.
C:\Users\majkinetor\AppData\Local\Temp\chocolatey\virtualbox\5.1.10\Oracle_VM_VirtualBox_Extension_Pack-5.1.10.vbox-extpack
0%...10%...20%...30%...40%...50%...60%...70%...80%...90%...100%
Successfully installed "Oracle VM VirtualBox Extension Pack".
Adding to PATH if needed
virtualbox installed to 'C:\Program Files\Oracle\VirtualBox'
virtualbox registered as vbox
 The upgrade of virtualbox was successful.
  Software installed as 'EXE', install location is likely default.
```

**Comment by majkinetor** [comment]
I think now it behaves correctly in all scenarios. Take a final look.

@ferventcoder plase, take a look too.

**Pull request merged**